mmtk/src/api.rs

```diff
@@ -485,39 +485,173 @@ pub extern "C" fn mmtk_get_obj_size(obj: ObjectReference) -> usize {
     }
 }
 
+#[allow(unused_variables)]
+fn assert_is_object(object: ObjectReference) {
+    // The checks are quite expensive. Dont run it in normal builds.
+    const ASSERT_OBJECT: bool = false;
+    if ASSERT_OBJECT {
+        #[cfg(debug_assertions)]
+        {
+            use crate::object_model::{is_object_in_immixspace, is_object_in_los};
+            if !mmtk_object_is_managed_by_mmtk(object.to_raw_address().as_usize()) {
+                panic!("{} is not managed by MMTk", object);
+            }
+            if !is_object_in_immixspace(&object) && !is_object_in_los(&object) {
+                // We will use VO bit in the following check. But if the object is not in immix space or LOS, we cannot do the check.
+                return;
+            }
+            if !object
+                .to_raw_address()
+                .is_aligned_to(ObjectReference::ALIGNMENT)
+            {
+                panic!(
+                    "{} is not aligned, it cannot be an object reference",
+                    object
+                )
+            }
+            if memory_manager::is_mmtk_object(object.to_raw_address()).is_none() {
+                error!("{} is not an object", object);
+                if let Some(base_ref) = memory_manager::find_object_from_internal_pointer(
+                    object.to_raw_address(),
+                    usize::MAX,
+                ) {
+                    panic!("{} is an internal pointer of {}", object, base_ref);
+                } else {
+                    panic!(
+                        "{} is not recognised as an object reference, or an internal reference",
+                        object
+                    );
+                }
+            }
+        }
+    }
+}
 #[no_mangle]
 pub extern "C" fn mmtk_pin_object(object: ObjectReference) -> bool {
+    assert_is_object(object);
     crate::early_return_for_non_moving_build!(false);
+    memory_manager::pin_object(object)
+}
+
+#[no_mangle]
+pub extern "C" fn mmtk_unpin_object(object: ObjectReference) -> bool {
+    assert_is_object(object);
+    crate::early_return_for_non_moving_build!(false);
+    memory_manager::unpin_object(object)
+}
+
+#[no_mangle]
+pub extern "C" fn mmtk_is_object_pinned(object: ObjectReference) -> bool {
+    assert_is_object(object);
+    crate::early_return_for_non_moving_build!(false);
+
+    memory_manager::is_pinned(object)
+}
 
-    // We may in the future replace this with a check for the immix space (bound check), which should be much cheaper.
-    if mmtk_object_is_managed_by_mmtk(object.to_raw_address().as_usize()) {
-        memory_manager::pin_object(object)
+#[no_mangle]
+pub extern "C" fn mmtk_pin_pointer(addr: Address) -> bool {
+    crate::early_return_for_non_moving_build!(false);
+    if mmtk_object_is_managed_by_mmtk(addr.as_usize()) {
+        if !crate::object_model::is_addr_in_immixspace(addr) {
```

Comment:
You only need to pin objects in the immixspace, correct? Wouldn't be enough to do a single check `if crate::object_model::is_addr_in_immixspace(addr)` instead of having the check `mmtk_object_is_managed_by_mmtk(addr.as_usize())` ?

Reply:
Yes. It should be fine. I made the change.